File: backend/scrape.jl
using HTTP
using Gumbo
using Cascadia
include("settings.jl")
include("gpt.jl")

function scrape_reviews(id::String, settings::Settings)::Tuple{Vector{String}, String}
    url = settings(id)
    @info "start get request for $url"

    response = HTTP.get(url; readtimeout=10)
    @info "finished response"
    reviews = String[]
    if response.status == 200
        @info "response status ok, start parsing"
        body = String(response.body)
        doc = Gumbo.parsehtml(body)
        selector = Selector(settings.element)
        matched_elements = eachmatch(selector, doc.root)
        titles = eachmatch(Selector(settings.title), doc.root)
        isempty(titles) ? t = "0" : t = replace(nodeText(titles[1]), r" |:|'" => "")
        for el in matched_elements
            txt = nodeText(el)
            push!(reviews, txt)
        end
    end
    @info "finished scraping $id with $(length(reviews)) reviews"
    reviews, t
end


function retry_if_empty(f::Function, args...; kwargs...)
    res, t = f(args...; kwargs...)
    while isempty(res)
        @info "Empty result, retrying..."
        sleep(3)
        res, t = f(args...; kwargs...)
    end
    return res, t
end

function chunk_reviews(reviews::Vector{String}, settings::Settings)::Vector{String}
    chunk_size = settings.chunksize
    chunks = String[]
    current_chunk = ""
    current_word_count = 0
    SMALLER = true

    for review in reviews
        review_word_count = length(split(review))
        if current_word_count + review_word_count <= chunk_size
            current_chunk = current_chunk * "\n next review: " * review
            current_word_count += review_word_count
            SMALLER = true
        else
            push!(chunks, current_chunk)
            current_chunk = review
            current_word_count = review_word_count
            SMALLER = false
        end
    end
    SMALLER ? push!(chunks, current_chunk) : nothing
    return chunks
end

function save_reviews_to_file(id, chunks, settings)
    for (index, chunk) in enumerate(chunks)
        fname = settings.datadir * "/review_$(id)_$(index).txt"
        @info "Writing $fname to disk"
        open(fname, "w") do io
            write(io, join(chunk, "\n\n"))
        end
    end
end

function save_summary_to_file(summary, title, settings)
    if !isdir(settings.datadir)
        @info "$(settings.datadir) does not exist in $pwd(), creating..."
        mkdir(settings.datadir)
    end
    fname = joinpath(settings.datadir, "summary_$(title).txt")
    @info "Writing $fname to disk"
    
    open(fname, "w") do io
        write(io, summary)
    end
end


function main(id)
    reviews, title = retry_if_empty(scrape_reviews, String(id), settings)
    chunks = chunk_reviews(reviews, settings)
    summary = loop_chunks(chunks, modelsettings)
    "--save" in ARGS ? save_summary_to_file(summary, title, settings) : @info "Add --save to save summary to disk"
    return summary
end


# @get "/scrape/{id}" main
# @get "/health" () -> "running!"
# serve(host="0.0.0.0")

File: python-backend/scrape.py
import time
import re
import requests
from bs4 import BeautifulSoup
from dataclasses import dataclass
from typing import List, Tuple
import tiktoken

@dataclass
class Settings:
    url: str
    element: str
    title: str
    timeout: int
    datadir: str
    chunksize: int
    model: str

    def __call__(self, id: str) -> str:
        return self.url.replace("id", str(id))




def scrape_reviews(id: str, settings: Settings) -> Tuple[List[str], str]:
    url = settings(id)
    print(f"Start get request for {url}")

    response = requests.get(url, timeout=settings.timeout)
    print("Finished response")
    reviews = []
    if response.status_code == 200:
        print("Response status ok, start parsing")
        soup = BeautifulSoup(response.text, 'html.parser')
        matched_elements = soup.select(settings.element)
        titles = soup.select(settings.title)
        t = "0" if not titles else re.sub(r" |:|'", "", titles[0].text)
        for el in matched_elements:
            reviews.append(el.text)
    print(f"Finished scraping {id} with {len(reviews)} reviews")
    return reviews, t

def retry_if_empty(f, *args, **kwargs):
    res, t = f(*args, **kwargs)
    while not res:
        print("Empty result, retrying...")
        time.sleep(3)
        res, t = f(*args, **kwargs)
    return res, t

def strip_tokens(review: str, num_tokens: int, max_review_tokens: int) -> str:
    ratio = max_review_tokens / num_tokens
    trim_to_length = int(len(review) * ratio)
    return review[:trim_to_length]

def trim_review(reviews: List[str], max_review_tokens: int = 1024, model_name: str ="gpt-4") -> List[Dict]:
    encoding = tiktoken.encoding_for_model(model_name) 
    chunks = [] 
    
    for review in reviews:
        num_tokens = len(encoding.encode(review))
        if num_tokens > max_review_tokens:
            review = strip_tokens(review, num_tokens, max_review_tokens)
            num_tokens = max_review_tokens
        
        chunks.append({"content" : review, "tokencount": num_tokens})
    chunks.sort(key=lambda chunk: chunk["tokencount"], reverse=True)
    return chunks

def main(id: str, settings: Settings) -> Tuple[List[Dict], str]:
    reviews, title = retry_if_empty(scrape_reviews, id, settings)
    chunks = trim_review(reviews)
    return chunks, title
        

